### PR TITLE
Migrating track length code from TOFEstimators into separate processor

### PR DESCRIPTION
**Pull request by dudarboh**
BEGINRELEASENOTES

- Migrated track length and  mean harmonic momentum code into separate processor to save CPU computing time

ENDRELEASENOTES

TODOs:

- [x] Finish migrating and cleaning up all relevant code to the TrackLengthProcessor
- [x] Do clean up of the TOFEstimators leaving only TOF related methods and output
- [x] Test the results with some example script, which uses both processors `TOFEstimators` and `TrackLengthProcessor` to calculate the mass distributions.
- [x] Update the documentation of both processors accordingly

### CODE REVIEW

**Comment by dudarboh** [comment]
I had a small test script to plot mass distribution.

I used 400 events from a single file:
`/pnfs/desy.de/ilc/prod/ilc/mc-2020/ild/rec/250-SetA/2f_hadronic_eL_pR/ILD_l5_o1_v02/v02-02/00015161/000/rv02-02.sv02-02.mILD_l5_o1_v02.E250-SetA.I500010.P2f_z_h.eL.pR.n001_209.d_rec_00015161_259.slcio`

Results before/after splitting the code look identical as it should be.

(sorry for slightly different image resolution)...

Code of trackLength and time-of-flight were already very well separated in different methods of `TOFEstimators`, so I it was very easy to copy whole methods without changing any logic inside so it is hard to mess something up here..

**Comment by dudarboh** [comment]
Concerning doxygen documentation:

`MarlinReco/doc/Doxyfile` looks very obsolete and does not include more than half of the processors inside `MarlinReco`including `TOFEstimators` and `TrackLength` now:
https://github.com/iLCSoft/MarlinReco/blob/a5ab3762ce8993c1631cdcecd1665503271c7979/doc/Doxyfile#L358

Full file list from iLCSoftDoc web page looks like this:
image

I would happily update this doxygen if there is a green light...

**Comment by tmadlener** [comment]
Hi @dudarboh, please go ahead with the doxygen fixes. If it is easily possible, could you put them into a separate PR to not let this one get even larger?

File: TimeOfFlight/README.md [hunk 1]
```diff
@@ -1,174 +1,142 @@
...
-## Extracting output parameters
+## Analysis
 
-Here is the pseudocode example on how one could extract PIDHandler parameters of *MyTofClosest0ps* results from the slcio file in your own Marlin processor and then use them to e.g. reconstruct the mass:
+After you run **TOFEstimators** and **TrackLength** processors you might want to run you analysis processor to e.g. calculate the mass of particles using time-of-flight information.
 
+Here is the code example how to do that:
 
-    void YourAmazingProcessor::processEvent(LCEvent* event){
+    float YourAmazingAnalysisProcessor::getParameterFromPID(ReconstructedParticle* pfo, PIDHandler& pidHandler, std::string algorithmName, std::string parameterName){
```
Comment (tmadlener) [inline TimeOfFlight/README.md:120]: You can make github highlight your code blocks in markdown via:
````
```cpp
// your code
```
````
https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

File: Tracking/TrackLength/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
+## Description
+
+TrackLength is the Marlin processor that calculates track length and square root of the harmonic mean of the squared momentum of charged particles.
+
+The detailed description of the source code is available in the [doxygen documentation](https://www.desy.de/~dudarboh/track_length_doc/html/index.html). <br>
```
Comment (tmadlener) [inline Tracking/TrackLength/README.md:5]: Once the doxygen generation is fixed, this should probably point to a more central location.

File: Tracking/TrackLength/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
+## Description
+
+TrackLength is the Marlin processor that calculates track length and square root of the harmonic mean of the squared momentum of charged particles.
```
Comment (tmadlener) [inline Tracking/TrackLength/README.md:3]: Maybe add one sentence here, that says that it was previously a part of TOFEstimators.

File: Tracking/TrackLength/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+      [00000071]                        . . .
+                 |      0 | 0.0000e+00 | 000000 |        9 | [ trackLengthToSET : 2.62e+03, trackLengthToEcal : 2.85e+03, momentumHMToSET : 3.84e+00, momentumHMToEcal : 3.84e+00,]
+
+## Analysis
+
+After you run **TOFEstimators** and **TrackLength** processors you might want to run you analysis processor to e.g. calculate the mass of particles using time-of-flight information.
+
+Here is the code example how to do that:
+
+    float YourAmazingAnalysisProcessor::getParameterFromPID(ReconstructedParticle* pfo, PIDHandler& pidHandler, std::string algorithmName, std::string parameterName){
```
Comment (tmadlener) [inline Tracking/TrackLength/README.md:101]: Also here, you could add some code highlighting.

**Review by tmadlener — COMMENTED** [review]
This looks pretty good to me from technical point of view already, as it is really mainly moving code from one part to another (or did I miss something substantial?). I think splitting this functionality into two processors makes sense, and it actually makes them easier to grasp. I have only very minor comments inline.

As already briefly discussed during the meeting today, the biggest point to solve is probably how to introduce this into a release without creating too many surprises for other users. Since the TOFEstimators are currently run in the standard reconstruction under the [HighLevelReco](https://github.com/iLCSoft/ILDConfig/blob/master/StandardConfig/production/HighLevelReco/HighLevelReco.xml#L308) section, could you prepare a pull request for ILDConfig with the necessary changes, to make that run with these changes?

From an analysis users perspective, I suppose there are also some (minor) code changes necessary to get the information out again? I think they are documented quite nicely here already, but we probably point other users to this documentation when we introduce these changes.

**Comment by dudarboh** [comment]
"... moving code from one part to another (or did I miss something substantial?)"

You didn't miss anything. It is indeed only moving the code.

And thanks for useful highlight tips :)

**Pull request merged**